[Opcache] Add Opcache\Jit attribute example

beberlei opened the pull request:
Changes:

```php
/** @jit */
function foo() {}
```

To:

```php
use Opcache\Jit;

<<Jit>>
function foo() {}
```

In addition allows disabling the JIT for a function regardless of the selected JIT trigger:

```php
use Opcache\Jit;

<<Jit(false)>>
function foo() {} // this never gets jitted
```
Also changes the precedence of an attribute to be always higher than any other trigger definition, which means with `<<Jit(false)>>` you can prevent jitting even for the "on script load" or "on first execute" trigger and with `<<Jit(true)>>` you can always force jitting a function or method on profiling trigger modes.

kooldev commented on ext/opcache/jit/zend_jit.c:
```diff
@@ -3135,20 +3135,8 @@ static int zend_jit_setup_hot_counters(zend_op_array *op_array)
-	}
-	return 0;
+	return op_array->attributes &&
+        zend_hash_str_exists(op_array->attributes, "opcache\\jit", sizeof("opcache\\jit")-1);
```
Can be simplified to `zend_hash_str_exists(op_array->attributes, ZEND_STRL("opcache\\jit");`

beberlei committed:
Add Attributes based on attributes_v2 RFC.

Co-authored-by: Martin Schröder <m.schroeder2007@gmail.com>

beberlei committed:
Update tests with more complex examples and edge cases.

beberlei committed:
Rename ReflectionAttribute::getAsObject to ReflectionAttribute::newInstance

kooldev committed:
Fixed attribute class validation bug.

kooldev committed:
Implemented many of the smaller changes suggested by Nikita Popov.

kooldev committed:
Introduced struct zend_attribute. Defer ASt evaluation to access time.

kooldev committed:
Added additional tests.

kooldev committed:
Implemented Opcache (memory-only) for new attribute structs.

beberlei committed:
Merge pull request #8 from koolkode/FixAttributeValidation

Implement Changes Requested in PR Review

kooldev committed:
Implemented attribute storage in file-based Opcache.

kooldev committed:
Fixed parameter offset bug. Refined test case to detect this.

beberlei committed:
Merge pull request #9 from koolkode/AttributeOpcache

Implement Opcache File Backend

beberlei committed:
Improve name resolution test to show all potential cases.

kooldev committed:
Enable external linkage. Fixed comments.

kooldev committed:
Implemented AST export of attribute nodes.

kooldev committed:
Removed ZEND_API from register ce function.

beberlei committed:
Merge pull request #10 from koolkode/PrintAst

AST Export / External Linkage

beberlei committed:
[jit] Introduce Opcache\Jit attribute to replace @jit

Instead of @jit in the docblock you can now use the <<Opcache\Jit>>
attribute. In addition this attribute allows you to disable the JIT
for a specific function as well with <<Opcache\Jit(false)>>.

Another behavior change is that the Opcache\Jit attribute has precedence
over any other trigger as well. This means you can allow to force JIT
in profiling triggers or disable JIT for a function regardless of trigger.

beberlei commented on ext/opcache/jit/zend_jit.c:
```diff
@@ -3294,12 +3320,6 @@ ZEND_EXT_API int zend_jit_op_array(zend_op_array *op_array, zend_script *script)
 		return zend_jit_setup_hot_trace_counters(op_array);
 	} else if (zend_jit_trigger == ZEND_JIT_ON_SCRIPT_LOAD) {
 		return zend_real_jit_func(op_array, script, NULL);
-	} else if (zend_jit_trigger == ZEND_JIT_ON_DOC_COMMENT) {
```
Having a Jit attribute on the op_array should take precedence over all the other triggers. If you configure the ON_ATTRIBUTE trigger, then it **only** jits on attributes.
